zeppelin-zengine/src/main/java/org/apache/zeppelin/scheduler/pool/DynamicThreadPool.java

```diff
@@ -0,0 +1,248 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zeppelin.scheduler.pool;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import org.quartz.SchedulerConfigException;
+import org.quartz.spi.ThreadPool;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This is a simple implementation of a Quartz thread pool.
+ *
+ * <CODE>Runnable</CODE> objects are sent to the pool by
+ * <code>{@link ThreadPoolExecutor}</code> with the
+ * <code>{@link #runInThread(Runnable)}</code> method,
+ * which blocks until a <code>Thread</code> becomes available.
+ *
+ * Pool based on <code>{@link SynchronousQueue}</code>.
+ *
+ * The pool has a dynamic number of <code>Thread</code>s, and grows or shrinks based on demand. It
+ * could be changed by calling <code>{@link #setThreadCount(int)}</code>
+ *
+ * Instance could be obtained using <code>{@link DynamicThreadPool#getInstance(String)}</code>
+ */
+public class DynamicThreadPool implements ThreadPool {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DynamicThreadPool.class);
+
+  private static final Map<String, DynamicThreadPool> INSTANCES = new ConcurrentHashMap<>();
+  private final Object poolEvent = new Object();
+
+  private final ThreadPoolExecutor executor =
+      new ThreadPoolExecutor(1, 10, 1, TimeUnit.MINUTES, new SynchronousQueue<>());
+
+  private volatile int priority = Thread.NORM_PRIORITY;
+
+  private String threadNamePrefix;
+  private String schedulerInstanceId;
+  private String schedulerInstanceName;
+
+  public static DynamicThreadPool getInstance(String schedulerInstanceId) {
+    return INSTANCES.get(schedulerInstanceId);
+  }
+
+  @Override
+  public boolean runInThread(Runnable runnable) {
+    if (runnable == null) {
+      return false;
+    }
+    try {
+      if (blockForAvailableThreads() > 0) {
+        executor.execute(new NotifableJob(runnable));
+      }
+      return true;
+    } catch (RejectedExecutionException e) {
+      LOGGER.error(e.toString());
```

Comment:
```suggestion
      LOGGER.error("Cron job rejected by thread pool", e);
```
Чаще всего, не стоит в логгер передавать `toString`. В крайнем случае если нечего сказать в сообщении и нужен только Exception, то
```java
LOGGER.error("", e);
```